packages/core/src/booter.ts

```diff
@@ -0,0 +1,83 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/core
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor} from '@loopback/context';
+
+/**
+ * Defines the requirements to implement a Booter for LoopBack applications:
+ * configure() : Promise<void>
+ * discover() : Promise<void>
+ * load(): Promise<void>
+ *
+ * A Booter will run through the above methods in order.
+ *
+ * @export
+ * @interface Booter
+ */
+export interface Booter {
+  /**
+   * Configure phase of the Booter. It should set options / defaults in this phase.
+   */
+  configure?(): Promise<void>;
+  /**
+   * Discover phase of the Booter. It should search for artifacts in this phase.
+   */
+  discover?(): Promise<void>;
+  /**
+   * Load phase of the Booter. It should bind the artifacts in this phase.
+   */
+  load?(): Promise<void>;
+}
+
+/**
+ * Export of an array of all the Booter phases supported by the interface
+ * above, in the order they should be run.
+ */
+export const BOOTER_PHASES = ['configure', 'discover', 'load'];
+
+/**
+ * Type Object for Options passed into .boot()
+ *
+ * @property projectRoot Root of project. All other artifacts are resolved relative to this
+ * @property booters An array of booters to bind to the application before running bootstrapper
+ * @property filter.booters An array of booters that should be run by the bootstrapper
+ * @property filter.phases An array of phases that should be run
+ */
+export type BootOptions = {
+  /**
+   * Root of the project. All other artifacts are resolved relative to this.
+   */
+  projectRoot: string;
+  /**
+   * Additional Properties
+   */
+  // tslint:disable-next-line:no-any
+  [prop: string]: any;
+};
+
+export type BootExecutionOptions = {
+  /**
+   * Optional array of Booter Classes to bind to the application before running bootstrapper.
+   */
+  booters?: Constructor<Booter>[];
+  /**
+   * Filter Object for Bootstrapper
+   */
+  filter?: {
+    /**
+     * Names of booters that should be run by Bootstrapper
+     */
+    booters?: string[];
```

Comment:
I have troubles imagining how is this API intended to be consumes.

Let's take API Connect for an example. When starting an app, how is API Connect supposed to know what booters the application uses under the hood? Do we have an easy-to-use API for obtaining that information? What criteria are app consumers going to apply when deciding which booters to include and which not?

I am proposing to leave this part out of scope of this pull request, so that we can land the initial implementation sooner and then have more time to discuss this additional feature as needed.

Reply:
I'm ok to drop `execOptions` entirely from the scope of this PR. 

---

But to answer the above question for future reference:

This api / option is intended to be used with API Connect / other service providing it's custom booter for example and providing only that name in this array so other booters are not run. Another option might be to want to selectively run a Booter from the DefaultList ... for example you may want to only run `ControllerBooter` to generate the OpenAPI Spec ... and this enables that so other Booters aren't run if not needed.